File: android/src/main/java/com/khanhtran/zalologin/ZaloLogin.java
package com.khanhtran.zalologin;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.util.Base64;
import android.widget.Toast;
//import android.widget.Toast;

import com.getcapacitor.JSObject;
import com.getcapacitor.NativePlugin;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.zing.zalo.zalosdk.kotlin.oauth.IAuthenticateCompleteListener;
import com.zing.zalo.zalosdk.kotlin.oauth.LoginVia;
import com.zing.zalo.zalosdk.kotlin.oauth.ZaloSDK;
import com.zing.zalo.zalosdk.kotlin.oauth.model.ErrorResponse;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONException;
import org.json.JSONObject;

import java.security.MessageDigest;
import java.util.Map;

@NativePlugin()
public class ZaloLogin extends Plugin   {
    public void load() {
        // Called when the plugin is first constructed in the bridge
//        ZaloSDK sdk = new ZaloSDK(getActivity());
    }

    @PluginMethod()
    public void echo(final PluginCall call) throws Exception {
//        Toast.makeText(getActivity(), "PRE CALL ECHO", Toast.LENGTH_LONG);
        saveCall(call);
        final Context ctx = getContext();
        new ZaloSDK(getContext()).authenticate(getActivity(), LoginVia.WEB, new IAuthenticateCompleteListener() {
            @Override
            public void onAuthenticateSuccess(long l, @NotNull String s, @NotNull Map<String, ?> map) {
                android.widget.Toast.makeText(ctx, "hello my friend", Toast.LENGTH_LONG).show();
                PluginCall lastCall = getSavedCall();
                JSObject data = new JSObject();
                data.put("hello", "ok");
                lastCall.resolve(data);
                call.resolve(data);
            }

            @Override
            public void onAuthenticateError(int i, @NotNull String s) {
                android.widget.Toast.makeText(ctx, "hello my friend", Toast.LENGTH_LONG).show();
                PluginCall lastCall = getSavedCall();
                JSObject data = new JSObject();
                data.put("hello", "ok");
                lastCall.resolve(data);
                call.resolve(data);
            }

            @Override
            public void onAuthenticateError(int i, @Nullable String s, @NotNull ErrorResponse errorResponse) {
                android.widget.Toast.makeText(ctx, "hello my friend", Toast.LENGTH_LONG).show();
                PluginCall lastCall = getSavedCall();
                JSObject data = new JSObject();
                data.put("hello", "ok");
                lastCall.resolve(data);
                call.resolve(data);
            }
        });
    }


    public void getApplicationHashKey() throws Exception {
        PackageInfo info = getContext().getPackageManager().getPackageInfo(getContext().getPackageName(), PackageManager.GET_SIGNATURES);
        for (Signature signature : info.signatures) {
            MessageDigest md = MessageDigest.getInstance("SHA");
            md.update(signature.toByteArray());
            String sig = Base64.encodeToString(md.digest(), Base64.DEFAULT).trim();
            if (sig.trim().length() > 0) {
//                Toast.makeText(getContext(), sig, Toast.LENGTH_LONG);
            }
        }
    }

}

File: android/src/main/java/com/khanhtran/zalologin/ZaloLogin.kt
package com.khanhtran.zalologin

import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.text.TextUtils
import android.util.Base64
import android.widget.Toast
import com.getcapacitor.*
import com.zing.zalo.zalosdk.kotlin.core.module.ZaloSDKInitProvider
import com.zing.zalo.zalosdk.kotlin.oauth.Constant
import com.zing.zalo.zalosdk.kotlin.oauth.IAuthenticateCompleteListener
import com.zing.zalo.zalosdk.kotlin.oauth.LoginVia
import com.zing.zalo.zalosdk.kotlin.oauth.ZaloSDK
import com.zing.zalo.zalosdk.kotlin.oauth.model.ErrorResponse
import java.lang.Exception
import java.security.MessageDigest

@NativePlugin(requestCodes = [Constant.ZALO_AUTHENTICATE_REQUEST_CODE])
class ZaloLogin : Plugin() {
    private lateinit var zaloSDK: ZaloSDK

    override fun load() {
        zaloSDK = ZaloSDK(activity);
    }

    @PluginMethod
    fun login(call: PluginCall) {
        val viaString = call.getString("loginVia", "APP_OR_WEB");
        val viaEnum: LoginVia;
        when (viaString) {
            "APP_OR_WEB" -> viaEnum = LoginVia.APP_OR_WEB;
            "APP" -> viaEnum = LoginVia.APP;
            "WEB" -> viaEnum = LoginVia.WEB;
            else -> { // Note the block
                throw Exception("loginVia is not valid");
            }
        }
        saveCall(call);
        zaloSDK.authenticate(activity, viaEnum, authenticateListener);
    }

    private val authenticateListener = object : IAuthenticateCompleteListener {
        override fun onAuthenticateSuccess(uid: Long, code: String, data: Map<String, Any>) {
            if (savedLastCall == null) {
                return;
            }
            val jsData = JSObject();
            jsData.put("displayName", data[Constant.user.DISPLAY_NAME]);
            jsData.put("uid", uid);
            jsData.put("code", code);
            savedLastCall.resolve(jsData);
        }

        override fun onAuthenticateError(errorCode: Int, message: String) {
            if (savedLastCall == null) {
                return;
            }
            savedLastCall.reject(message);
        }

        override fun onAuthenticateError(errorCode: Int, errorMsg: String?, errorResponse: ErrorResponse) {
            if (savedLastCall == null) {
                return;
            }
            savedLastCall.reject(errorMsg);
        }
    }

    @PluginMethod
    fun getApplicationHashKey(call: PluginCall) {
        val info = context.packageManager.getPackageInfo(context.packageName, PackageManager.GET_SIGNATURES)
        for (signature in info.signatures) {
            val md = MessageDigest.getInstance("SHA")
            md.update(signature.toByteArray())
            val sig = Base64.encodeToString(md.digest(), Base64.DEFAULT).trim { it <= ' ' }
            if (sig.trim { it <= ' ' }.length > 0) {
                val data = JSObject();
                data.put("signature", sig);
                call.resolve(data);
            }
        }
    }

    override fun handleOnActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        zaloSDK.onActivityResult(activity, requestCode, resultCode, data);
    }
}